test/integration/cdi/example-driver/kubeletplugin/kubeletplugin.go

```diff
@@ -0,0 +1,135 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	cdiapi "github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	specs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"k8s.io/klog/v2"
+	drapbv1 "k8s.io/kubernetes/pkg/kubelet/apis/dra/v1alpha1"
+	server "k8s.io/kubernetes/test/integration/cdi/example-driver/kubeletplugin/server"
+)
+
+type PluginConfig struct {
+	DriverName             string
+	DraAddress             string
+	PluginRegistrationPath string
+}
+
+type examplePlugin struct {
+	config PluginConfig
+}
+
+// newExamplePlugin returns an initialized examplePlugin instance
+func newExampleDriver(config PluginConfig) (*examplePlugin, error) {
+
+	if err := os.MkdirAll(filepath.Dir(config.DraAddress), 0750); err != nil {
+		return nil, err
+	}
+
+	return (&examplePlugin{
+		config: config,
+	}), nil
+}
+
+func StartPlugin(config *PluginConfig) error {
+
+	err := server.StartRegistrar(config.DriverName, config.DraAddress, config.PluginRegistrationPath)
+	if err != nil {
+		return err
+	}
+
+	driver, err := newExampleDriver(*config)
+	if err != nil {
+		return err
+	}
+
+	s := server.NewNonBlockingGRPCServer()
+	// ex itself implements NodeServer.
+	s.Start(config.DraAddress, driver)
+	s.Wait()
+
+	return nil
+}
+
+var jsonFilePath = "/etc/cdi/vendor.json"
```

Comment:
You cannot hard-code a file name like this.

@bart0sh: is there a naming convention that vendors should follow to avoid overwriting each others files?

Reply:
I got this when I deployed shared-resourceclaim.yaml.
Interestingly, the claimName are actually unique.
Do you think, we can use this for naming the json files?

I0805 02:00:47.574983 2621373 kubeletplugin.go:86] NodePrepareResource is called: request: &NodePrepareResourceRequest{**Namespace:default,ClaimUid:094e7ec3-4713-4404-97b3-08ddd180a338,ClaimName:f1f5c13c-e908-4edf-9fbc-71ccbc643995-example**,ResourceHandle:{"user_a":"b"},}
I0805 02:00:47.575035 2621373 kubeletplugin.go:94] Creating CDI Files
I0805 02:00:47.575060 2621373 kubeletplugin.go:135] CDI spec is validated
I0805 02:00:47.575197 2621373 kubeletplugin.go:147] Json file is created at /etc/cdi/vendor.json
I0805 02:00:47.575217 2621373 kubeletplugin.go:152] Return: vendor.com/device=example
**I0805 02:00:47.580004 2621373 kubeletplugin.go:86] NodePrepareResource is called: request: &NodePrepareResourceRequest{**Namespace:default,ClaimUid:094e7ec3-4713-4404-97b3-08ddd180a338,ClaimName:441e23ec-9cac-47d1-898b-b1e879d68646-example**,ResourceHandle:{"user_a":"b"},}**
I0805 02:00:47.580055 2621373 kubeletplugin.go:94] Creating CDI Files
I0805 02:00:47.580090 2621373 kubeletplugin.go:135] CDI spec is validated
I0805 02:00:47.580274 2621373 kubeletplugin.go:147] Json file is created at /etc/cdi/vendor.json
I0805 02:00:47.580298 2621373 kubeletplugin.go:152] Return: vendor.com/device=example

Reply:
Maybe done.! If you think it's done. Please, resolve it.! I am not sure.

Reply:
@bart0sh @pohly Please see this log and let me know what to do. Idk why uid is same, and name is unique.

Reply:
@hj-johannes-lee uid is the same because 2 pods share the same claim.

Claim name is not passed to this call. It's a bug in my code. Here is [the fix](https://github.com/pohly/kubernetes/pull/8)

Here is a plugin output with the applied fix:
```
I0806 02:16:14.676485    3776 nodeserver.go:51] NodePrepareResource is called: request: &NodePrepareResourceRequest{Namespace:default,ClaimUid:61e69ce7-e1f7-4d15-9f6f-7c6192ba7a91,ClaimName:shared-claim,ResourceHandle:{"user_a":"b"},}
I0806 02:16:14.676552    3776 nodeserver.go:59] Creating CDI Files
I0806 02:16:14.676575    3776 nodeserver.go:89] CDI spec is validated
I0806 02:16:14.676979    3776 nodeserver.go:101] Json file is created at /etc/cdi/vendor.json
I0806 02:16:14.677001    3776 nodeserver.go:106] Return: vendor.com/device=example
I0806 02:16:14.684296    3776 nodeserver.go:51] NodePrepareResource is called: request: &NodePrepareResourceRequest{Namespace:default,ClaimUid:61e69ce7-e1f7-4d15-9f6f-7c6192ba7a91,ClaimName:shared-claim,ResourceHandle:{"user_a":"b"},}
I0806 02:16:14.684429    3776 nodeserver.go:59] Creating CDI Files
I0806 02:16:14.684465    3776 nodeserver.go:89] CDI spec is validated
I0806 02:16:14.684733    3776 nodeserver.go:101] Json file is created at /etc/cdi/vendor.json
I0806 02:16:14.684756    3776 nodeserver.go:106] Return: vendor.com/device=example
I0806 02:16:18.556491    3776 nodeserver.go:112] NodeUnprepareResource is called: request: &NodeUnprepareResourceRequest{Namespace:default,ClaimUid:61e69ce7-e1f7-4d15-9f6f-7c6192ba7a91,ClaimName:shared-claim,CdiDevice:[vendor.com/device=example],}
I0806 02:16:19.561129    3776 nodeserver.go:112] NodeUnprepareResource is called: request: &NodeUnprepareResourceRequest{Namespace:default,ClaimUid:61e69ce7-e1f7-4d15-9f6f-7c6192ba7a91,ClaimName:shared-claim,CdiDevice:[vendor.com/device=example],}
```

Reply:
So, to make CDI file name unique you'd need to add random suffix or something similar I guess.

Reply:
Another idea is to make a unique hash out of parameters and use that as a suffix. This would complicate your code a little bit though as you should keep track of used file names and only delete them after the last NodeUnprepareResource call for the same resource.

Reply:
Why is NodePrepareResource called twice when it is the same claim? It should only be called once!

Reply:
@pohly should be fixed in #9

Reply:
Here is a plugin output fore the shared claim example with changes from #9:
```
I0809 00:31:42.190905   11002 server.go:106] Found KUBECONFIG environment variable set, using that..
I0809 00:31:42.194965   11002 noderegistrar.go:49] Starting Registration Server at: /var/lib/kubelet/plugins_registry/example-driver.cdi.k8s.io-reg.sock
I0809 00:31:42.195060   11002 noderegistrar.go:58] Registration Server started at: /var/lib/kubelet/plugins_registry/example-driver.cdi.k8s.io-reg.sock
I0809 00:31:43.710827   11002 registrationserver.go:31] Received GetInfo call: &InfoRequest{}
I0809 00:31:43.711683   11002 registrationserver.go:42] Received NotifyRegistrationStatus call: &RegistrationStatus{PluginRegistered:true,Error:,}
I0809 00:31:54.967270   11002 nodeserver.go:51] NodePrepareResource is called: request: &NodePrepareResourceRequest{Namespace:default,ClaimUid:0df67e02-93ff-4e14-8f00-c815af4f511f,ClaimName:shared-claim,ResourceHandle:{"user_a":"b"},}
I0809 00:31:54.967332   11002 nodeserver.go:59] Creating CDI Files
I0809 00:31:54.967363   11002 nodeserver.go:89] CDI spec is validated
I0809 00:31:54.967747   11002 nodeserver.go:101] Json file is created at /etc/cdi/vendor.json
I0809 00:31:54.967767   11002 nodeserver.go:106] Return: vendor.com/device=example
I0809 00:31:58.788329   11002 nodeserver.go:112] NodeUnprepareResource is called: request: &NodeUnprepareResourceRequest{Namespace:default,ClaimUid:0df67e02-93ff-4e14-8f00-c815af4f511f,ClaimName:shared-claim,CdiDevice:[vendor.com/device=example],}
```

Reply:
@bart0sh Thanks! :)